packages/wallet/test/services/SmartTxSubmitProvider.test.ts

```diff
@@ -0,0 +1,140 @@
+import { BehaviorSubject, EMPTY, of } from 'rxjs';
+import { Cardano, ProviderError, ProviderFailure, TxSubmitProvider, coreToCsl, util } from '@cardano-sdk/core';
+import { ConnectionStatus, SmartTxSubmitProvider, TipSlot } from '../../src';
+import { RetryBackoffConfig } from 'backoff-rxjs';
+import { flushPromises } from '@cardano-sdk/util-dev';
+import { mockTxSubmitProvider } from '../mocks';
+
+describe('SmartTxSubmitProvider', () => {
+  let underlyingProvider: jest.Mocked<TxSubmitProvider>;
+  let provider: SmartTxSubmitProvider;
+  const retryBackoffConfig: RetryBackoffConfig = { initialInterval: 1 };
+
+  beforeEach(() => (underlyingProvider = mockTxSubmitProvider()));
+
+  describe('healthCheck', () => {
+    it('calls underlying provider', () => {
+      provider = new SmartTxSubmitProvider(
+        { retryBackoffConfig },
+        { connectionStatus$: EMPTY, tip$: EMPTY, txSubmitProvider: underlyingProvider }
+      );
+      expect(provider.healthCheck()).toEqual(underlyingProvider.healthCheck());
+      expect(underlyingProvider.healthCheck).toBeCalledTimes(2);
+    });
+  });
+
+  describe('submitTx', () => {
+    const txWithoutValidityInterval: Cardano.NewTxAlonzo = {
+      body: {
+        fee: 0n,
+        inputs: [],
+        outputs: [],
+        validityInterval: {}
+      },
+      id: Cardano.TransactionId('6804edf9712d2b619edb6ac86861fe93a730693183a262b165fcc1ba1bc99cad'),
+      witness: { signatures: new Map() }
+    };
+    const validityInterval = { invalidBefore: 5, invalidHereafter: 10 };
+    const txWithoutValidityIntervalHex = util.bytesToHex(coreToCsl.tx(txWithoutValidityInterval).to_bytes());
+    const txWithValidityIntervalHex = util.bytesToHex(
+      coreToCsl
+        .tx({
+          ...txWithoutValidityInterval,
+          body: {
+            ...txWithoutValidityInterval.body,
+            validityInterval
+          }
+        })
+        .to_bytes()
+    );
+
+    describe('all preconditions are met', () => {
+      beforeEach(() => {
+        provider = new SmartTxSubmitProvider(
+          { retryBackoffConfig },
+          {
+            connectionStatus$: of(ConnectionStatus.up),
+            tip$: of({ slot: validityInterval.invalidBefore }),
+            txSubmitProvider: underlyingProvider
+          }
+        );
+      });
+
+      it('calls underlying provider when online and tx has no validity interval', async () => {
+        await provider.submitTx({ signedTransaction: txWithoutValidityIntervalHex });
+        expect(underlyingProvider.submitTx).toBeCalledTimes(1);
+      });
+
+      it('calls underlying provider when online and within validity interval', async () => {
+        await provider.submitTx({ signedTransaction: txWithValidityIntervalHex });
+        expect(underlyingProvider.submitTx).toBeCalledTimes(1);
+      });
+
+      it('re-submits transactions that failed to submit due to a connection or recoverable server issue', async () => {
```

Comment:
```suggestion
      it('re-submits transactions that failed to submit due to a connection or recoverable provider issue', async () => {
```
This could be used in Node.js too :sunglasses:

Reply:
3af7e3498ba6f0a20cc1fa52c410a4a7dec9ea0a